src/inputs/pcap/PcapInputStream.cpp

```diff
@@ -374,17 +392,26 @@ void PcapInputStream::process_raw_packet(pcpp::RawPacket *rawPacket)
 
     auto timestamp = rawPacket->getPacketTimeStamp();
     // interface to handlers
-    packet_signal(packet, dir, l3, l4, timestamp);
+    std::unique_lock lock(_input_mutex);
+    for (auto &callback : _callbacks) {
+        dynamic_cast<PcapInputStreamCallback *>(callback.get())->process_packet_cb(packet, dir, l3, l4, timestamp);
+    }
 
     if (l4 == pcpp::UDP) {
-        udp_signal(packet, dir, l3, pcpp::hash5Tuple(&packet), timestamp);
+        for (auto &callback : _callbacks) {
+            dynamic_cast<PcapInputStreamCallback *>(callback.get())->process_udp_packet_cb(packet, dir, l3, pcpp::hash5Tuple(&packet), timestamp);
+        }
     } else if (l4 == pcpp::TCP) {
+        lock.unlock();
```

Comment:
don't know about this part as well. need to release and lock the mutex again (line 411) to not have deadlock. could change it to recursive_mutex

Reply:
why do we need to lock and unlock, is this recursive?